### PR TITLE
Handle exceptions while running HTTP web server for remote.

### LINKED ISSUE #3121: Unhandled socket IO exception
From @smortaz 

```
Description: The process was terminated due to an unhandled exception.
Exception Info: System.Net.Sockets.SocketException
   at System.Net.Sockets.Socket.BeginSend(Byte[], Int32, Int32, System.Net.Sockets.SocketFlags, System.AsyncCallback, System.Object)
   at System.Net.Sockets.NetworkStream.BeginWrite(Byte[], Int32, Int32, System.AsyncCallback, System.Object)

Exception Info: System.IO.IOException
   at System.Net.Sockets.NetworkStream.BeginWrite(Byte[], Int32, Int32, System.AsyncCallback, System.Object)
   at System.Net.Security._SslStream.StartWriting(Byte[], Int32, Int32, System.Net.AsyncProtocolRequest)
   at System.Net.Security._SslStream.ProcessWrite(Byte[], Int32, Int32, System.Net.AsyncProtocolRequest)
   at System.Net.Security._SslStream.BeginWrite(Byte[], Int32, Int32, System.AsyncCallback, System.Object)
   at System.Net.TlsStream.BeginWrite(Byte[], Int32, Int32, System.AsyncCallback, System.Object)
   at System.Net.PooledStream.BeginWrite(Byte[], Int32, Int32, System.AsyncCallback, System.Object)
   at System.IO.Stream+<>c.<BeginEndWriteAsync>b__53_0(System.IO.Stream, ReadWriteParameters, System.AsyncCallback, System.Object)
   at System.Threading.Tasks.TaskFactory`1[[System.Threading.Tasks.VoidTaskResult, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].FromAsyncTrim[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.IO.Stream+ReadWriteParameters, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]](System.__Canon, ReadWriteParameters, System.Func`5<System.__Canon,ReadWriteParameters,System.AsyncCallback,System.Object,System.IAsyncResult>, System.Func`3<System.__Canon,System.IAsyncResult,System.Threading.Tasks.VoidTaskResult>)
   at System.IO.Stream.BeginEndWriteAsync(Byte[], Int32, Int32)
   at System.IO.Stream.WriteAsync(Byte[], Int32, Int32, System.Threading.CancellationToken)
   at System.Net.DelegatedStream.WriteAsync(Byte[], Int32, Int32, System.Threading.CancellationToken)
   at System.Net.WebSockets.WebSocketConnectionStream+WebSocketConnection.WriteAsync(Byte[], Int32, Int32, System.Threading.CancellationToken)
   at System.Net.DelegatedStream.WriteAsync(Byte[], Int32, Int32, System.Threading.CancellationToken)
   at System.Net.WebSockets.WebSocketConnectionStream.<>n__1(Byte[], Int32, Int32, System.Threading.CancellationToken)
   at System.Net.WebSockets.WebSocketConnectionStream+<WriteAsync>d__22.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
   at Microsoft.AspNetCore.WebSockets.Protocol.CommonWebSocket+<SendAsync>d__33.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
   at Microsoft.Common.Core.WebSocketHelper+<DoWebSocketReceiveSendAsync>d__0.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(System.Threading.Tasks.Task)
   at Microsoft.R.Host.Client.BrokerServices.RemoteUriWebService+<GetResponseAsync>d__4.MoveNext()
```

### PR DESCRIPTION
**Pull request by karthiknadig**
Fixes #3121

### CODE REVIEW

File: src/Host/Client/Impl/BrokerServices/RemoteUriWebService.cs [hunk 1]
```diff
@@ -73,17 +83,23 @@ public class RemoteUriWebService : IRemoteUriWebService {
                 }
             } catch (WebException wex) when (wex.Status == WebExceptionStatus.ProtocolError) {
                 response = wex.Response as HttpWebResponse;
+            } catch (TaskCanceledException) {
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/BrokerServices/RemoteUriWebService.cs]: `OperationCanceledException`

**Comment by karthiknadig** [comment]
ping.

File: src/Host/Client/Impl/BrokerServices/RemoteUriWebService.cs [hunk 2]
```diff
@@ -13,13 +13,22 @@
...
 using Microsoft.R.Host.Protocol;
 using static System.FormattableString;
 
 namespace Microsoft.R.Host.Client.BrokerServices {
     public class RemoteUriWebService : IRemoteUriWebService {
-        public RemoteUriWebService(string baseUri) {
+        private ICoreServices _services;
+        private IConsole _console;
+
+        private IActionLog Log => _services?.Log;
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/BrokerServices/RemoteUriWebService.cs]: I'd rather use a field and assign it in ctor.
Also, do we have a case when log is null?

File: src/Host/Client/Impl/BrokerServices/RemoteUriWebService.cs [hunk 3]
```diff
@@ -13,13 +13,23 @@
...
+        private ICoreServices _services;
+        private IConsole _console;
+
+        private readonly IActionLog Log;
+
+        public RemoteUriWebService(string baseUri, ICoreServices services, IConsole console) {
             PostUri = new Uri(new Uri(baseUri), new Uri("/remoteuri", UriKind.Relative));
+            _services = services;
+            _console = console;
+            Log = _services.Log;
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/BrokerServices/RemoteUriWebService.cs]: field name should be `_log`, and I think `_services` and `_console` can also be read-only.

**Comment by AlexanderSher** [comment]
👍

**Comment by MikhailArkhipov** [comment]
Needs to go to 1_0

**Comment by karthiknadig** [comment]
Ported to 1_0 e93c2a5

**Pull request merged**